File: bin/sectLabel/genericSectExtract.rb
#!/usr/bin/env ruby
require 'find'
# Emma

pwd = File.dirname(__FILE__)

@CRFPP  = ENV['CRFPP_HOME'] ? "#{ENV['CRFPP_HOME']}/bin" : "#{pwd}/../../crfpp"
@SRC    = "#{pwd}/genericSect"
@DATA   = "#{pwd}/../../resources/sectLabel/"
@TEST_DIR = "/tmp/"

name  = "#{Time.now.to_i}-#{Process.pid}"

cmd = "ruby #{@SRC}/extractFeature.rb #{ARGV[0]} > #{@TEST_DIR}/#{name}.test"
system(cmd)

cmd = "#{@CRFPP}/crf_test -m #{@DATA}/genericSect.model  #{@TEST_DIR}/#{name}.test >  #{@TEST_DIR}/#{name}.out"
system(cmd)


if ARGV[1] != nil
	g = File.open("#{ARGV[1]}", "w")

	cmd = "chmod 777 #{ARGV[1]}"
	system(cmd)
end

f = File.open("#{@TEST_DIR}/#{name}.out")
while !f.eof do
	str = f.gets.chomp.strip
	if str != ""
		l = str.split(" ")
		output = l.at(l.length-1)
		while output.index("-") != nil
			output = output.sub("-", " ")
		end
		if output == "related works"
			output = "related work"
		end
		if ARGV[1] == nil
			puts "#{output}"
		else
			g.write("#{output}\n")
		end
	end
end
f.close
if ARGV[1] != nil
	g.close
end

File.unlink("#{@TEST_DIR}/#{name}.out")
File.unlink("#{@TEST_DIR}/#{name}.test")

File: bin/sectLabel/genericSectExtract.pl
#!perl

use FindBin;
use lib "$FindBin::Bin/../lib";

use File::Spec;
use String::Strip;
use File::Temp;
use IPC::Run3;
use ParsCit;

my $crf_test = ParsCit->crf_test_path;
my $data = ParsCit->_resource_path('resources/sectLabel');
my $test_dir = File::Temp->newdir;

sub main {
	my $name  = "@{[ time ]}-$$";

	my $test_file = File::Spec->catfile($test_dir, "${name}.test" );
	open( my $extract_fh, '>:encoding(UTF-8)', $test_file );
	run3 [
		$^X,
		File::Spec->catfile($FindBin::Bin, qw(extractFeature.pl)),
		$ARGV[0],
	], \undef, $extract_fh;
	close $extract_fh;

	my $out_file =  File::Spec->catfile($test_dir, "${name}.out" );
	open( my $out_fh, '>:encoding(UTF-8)', $out_file );
	run3 [
		$crf_test,
		qw(-m), File::Spec->catfile($data, qw(genericSect.model)),
		$test_file,
	], \undef, $out_fh;
	close $out_fh;

	my $output_fh;
	if( $ARGV[1] ) {
		open( $output_fh, '>:encoding(UTF-8)', $ARGV[1] );
	} else {
		binmode STDOUT, ":encoding(UTF-8)";
		$output_fh = \*STDOUT;
	}

	open( my $out_read_fh, '<:encoding(UTF-8)', $out_file);
	while( my $str = <$out_read_fh> ) {
		chomp $str;
		StripLTSpace($str);
		next unless $str;

		my @words = split ' ', $str;
		my $output = $words[-1];
		$output =~ s/-/ /g;

		$output = 'related work' if $output eq 'related works';

		print $output_fh "$output\n";
	}
}

main;
